File: Sources/OpenRenderBox/Path/ORBPathCallbacks.cpp
//
//  ORBPathCallbacks.cpp
//  OpenRenderBox
//
//  Audited for 6.5.1
//  Status: Complete

#include <OpenRenderBox/ORBPathCallbacks.h>
#include <OpenRenderBoxCxx/Util/assert.hpp>

#if ORB_TARGET_OS_DARWIN

CG_EXTERN size_t CGPathGetNumberOfPoints(CGPathRef cg_nullable path);
CG_EXTERN size_t CGPathGetNumberOfElements(CGPathRef cg_nullable path);
typedef void (^CGPathApplyBlock2)(const CGPathElement * element, bool *stop);
CG_EXTERN void CGPathApplyWithBlock2(CGPathRef path, CGPathApplyBlock2 CF_NOESCAPE block);

namespace {
    uint32_t cgpath_bezier_order(CGPathRef cgPath) {
        size_t pointsCount = CGPathGetNumberOfPoints(cgPath);
        size_t elementsCount = CGPathGetNumberOfElements(cgPath);
        if (pointsCount > elementsCount) { return 3; }
        if (pointsCount != 3 || elementsCount != 3) { return 1; }
        bool hasCurve = false;
        CGPathApply(cgPath, &hasCurve, [](void *info, const CGPathElement *element) {
            if (element->type == kCGPathElementAddQuadCurveToPoint
                || element->type == kCGPathElementAddCurveToPoint) {
                *static_cast<bool*>(info) = true;
            }
        });
        return hasCurve ? 3 : 1;
    }
}

const ORBPathCallbacks ORBPathCGPathCallbacks = {
    .flags = {},
    .retain = CFRetain,
    .release = CFRelease,
    .apply = +[](const void *object, void *info, ORBPathApplyCallback callback) -> bool {
        CGPathRef cgPath = reinterpret_cast<CGPathRef>(object);
        __block bool shouldStop = false;
        CGPathApplyWithBlock2(cgPath, ^(const CGPathElement *element, bool *stop) {
            bool result = callback(info,
                                   (ORBPathElement)element->type,
                                   (CGFloat *)element->points,
                                   nullptr);
            if (!result) {
                *stop = true;
                shouldStop = true;
            }
        });
        return !shouldStop;
    },
    .isEqual = +[](const void *object, const void *otherObject) -> bool {
        return CGPathEqualToPath(static_cast<CGPathRef>(object), static_cast<CGPathRef>(otherObject));
    },
    .isEmpty = +[](const void *object) -> bool {
        return CGPathIsEmpty(static_cast<CGPathRef>(object));
    },
    .isSingleElement = +[](const void *object) -> bool {
        return false;
    },
    .bezierOrder = +[](const void *object) -> uint32_t {
        return cgpath_bezier_order(static_cast<CGPathRef>(object));
    },
    .boundingRect = +[](const void *object) -> CGRect {
        return CGPathGetPathBoundingBox(static_cast<CGPathRef>(object));
    },
    .cgPath = +[](const void *object) -> CGPathRef {
        return static_cast<CGPathRef>(object);
    },
    .next = nullptr,
};

#endif /* ORB_TARGET_OS_DARWIN */

File: Sources/OpenRenderBox/Path/ORBPathCallbacks.swift
//
//  ORBPathCallbacks.swift
//  OpenRenderBox

#if !OPENRENDERBOX_CF_CGTYPES
public import OpenCoreGraphicsShims

// MARK: - ORBPath.Callbacks

extension ORBPath {
    /// Callbacks structure for path operations
    /// This allows different path storage types (CGPath, custom storage, etc.) to provide their own implementations
    public struct Callbacks {
        /// Flags for path callbacks
        public struct Flags {
            public var unknown0: UInt8
            public var unknown1: UInt8
            public var isExtended: Bool
            public var padding: (UInt8, UInt8, UInt8, UInt8, UInt8)

            public init() {
                self.unknown0 = 0
                self.unknown1 = 0
                self.isExtended = false
                self.padding = (0, 0, 0, 0, 0)
            }

            public init(unknown0: UInt8, unknown1: UInt8, isExtended: Bool, padding: (UInt8, UInt8, UInt8, UInt8, UInt8)) {
                self.unknown0 = unknown0
                self.unknown1 = unknown1
                self.isExtended = isExtended
                self.padding = padding
            }
        }

        public var flags: ORBPath.Callbacks.Flags
        public var retain: ((UnsafeRawPointer) -> UnsafeRawPointer)?
        public var release: ((UnsafeRawPointer) -> Void)?
        public var apply: ((UnsafeRawPointer, UnsafeMutableRawPointer, ORBPathApplyCallback?) -> Bool)?
        public var isEqual: ((UnsafeRawPointer, UnsafeRawPointer) -> Bool)?
        public var isEmpty: ((UnsafeRawPointer) -> Bool)?
        public var isSingleElement: ((UnsafeRawPointer) -> Bool)?
        public var bezierOrder: ((UnsafeRawPointer) -> UInt32)?
        public var boundingRect: ((UnsafeRawPointer) -> CGRect)?
        public var cgPath: ((UnsafeRawPointer) -> Unmanaged<CGPath>?)?
        public var next: ((UnsafeRawPointer) -> UnsafePointer<ORBPath.Callbacks>?)?
        
        public init() {
            self.flags = ORBPath.Callbacks.Flags()
            self.retain = nil
            self.release = nil
            self.apply = nil
            self.isEqual = nil
            self.isEmpty = nil
            self.isSingleElement = nil
            self.bezierOrder = nil
            self.boundingRect = nil
            self.cgPath = nil
            self.next = nil
        }
        
        public init(
            flags: ORBPath.Callbacks.Flags,
            retain: ((UnsafeRawPointer) -> UnsafeRawPointer)?,
            release: ((UnsafeRawPointer) -> Void)?,
            apply: ((UnsafeRawPointer, UnsafeMutableRawPointer, ORBPathApplyCallback?) -> Bool)?,
            isEqual: ((UnsafeRawPointer, UnsafeRawPointer) -> Bool)?,
            isEmpty: ((UnsafeRawPointer) -> Bool)?,
            isSingleElement: ((UnsafeRawPointer) -> Bool)?,
            bezierOrder: ((UnsafeRawPointer) -> UInt32)?,
            boundingRect: ((UnsafeRawPointer) -> CGRect)?,
            cgPath: ((UnsafeRawPointer) -> Unmanaged<CGPath>?)?,
            next: ((UnsafeRawPointer) -> UnsafePointer<ORBPath.Callbacks>?)?
        ) {
            self.flags = flags
            self.retain = retain
            self.release = release
            self.apply = apply
            self.isEqual = isEqual
            self.isEmpty = isEmpty
            self.isSingleElement = isSingleElement
            self.bezierOrder = bezierOrder
            self.boundingRect = boundingRect
            self.cgPath = cgPath
            self.next = next
        }
    }
}

// MARK: - ORBPath.Callbacks Static Properties

extension ORBPath.Callbacks {
    /// Global callbacks for CGPath-backed paths
    public static var cgPath: ORBPath.Callbacks {
        _openRenderBoxUnimplementedFailure()
    }
}

// MARK: - ORBPath.CallbacksExtended

extension ORBPath {
    /// Extended callbacks structure with additional extended callbacks argument
    public struct CallbacksExtended {
        public var flags: ORBPath.Callbacks.Flags
        public var retain: ((UnsafeRawPointer) -> UnsafeRawPointer)?
        public var release: ((UnsafeRawPointer) -> Void)?
        public var apply: ((UnsafeRawPointer, UnsafeMutableRawPointer, ORBPathApplyCallback?, UnsafePointer<ORBPath.CallbacksExtended>) -> Bool)?
        public var isEqual: ((UnsafeRawPointer, UnsafeRawPointer, UnsafePointer<ORBPath.CallbacksExtended>) -> Bool)?
        public var isEmpty: ((UnsafeRawPointer, UnsafePointer<ORBPath.CallbacksExtended>) -> Bool)?
        public var isSingleElement: ((UnsafeRawPointer, UnsafePointer<ORBPath.CallbacksExtended>) -> Bool)?
        public var bezierOrder: ((UnsafeRawPointer, UnsafePointer<ORBPath.CallbacksExtended>) -> UInt32)?
        public var boundingRect: ((UnsafeRawPointer, UnsafePointer<ORBPath.CallbacksExtended>) -> CGRect)?
        public var cgPath: ((UnsafeRawPointer, UnsafePointer<ORBPath.CallbacksExtended>) -> Unmanaged<CGPath>?)?
        public var next: ((UnsafeRawPointer, UnsafePointer<ORBPath.CallbacksExtended>) -> UnsafePointer<ORBPath.CallbacksExtended>?)?
        
        public init() {
            self.flags = ORBPath.Callbacks.Flags()
            self.retain = nil
            self.release = nil
            self.apply = nil
            self.isEqual = nil
            self.isEmpty = nil
            self.isSingleElement = nil
            self.bezierOrder = nil
            self.boundingRect = nil
            self.cgPath = nil
            self.next = nil
        }

        public init(
            flags: ORBPath.Callbacks.Flags,
            retain: ((UnsafeRawPointer) -> UnsafeRawPointer)?,
            release: ((UnsafeRawPointer) -> Void)?,
            apply: ((UnsafeRawPointer, UnsafeMutableRawPointer, ORBPathApplyCallback?, UnsafePointer<ORBPath.CallbacksExtended>) -> Bool)?,
            isEqual: ((UnsafeRawPointer, UnsafeRawPointer, UnsafePointer<ORBPath.CallbacksExtended>) -> Bool)?,
            isEmpty: ((UnsafeRawPointer, UnsafePointer<ORBPath.CallbacksExtended>) -> Bool)?,
            isSingleElement: ((UnsafeRawPointer, UnsafePointer<ORBPath.CallbacksExtended>) -> Bool)?,
            bezierOrder: ((UnsafeRawPointer, UnsafePointer<ORBPath.CallbacksExtended>) -> UInt32)?,
            boundingRect: ((UnsafeRawPointer, UnsafePointer<ORBPath.CallbacksExtended>) -> CGRect)?,
            cgPath: ((UnsafeRawPointer, UnsafePointer<ORBPath.CallbacksExtended>) -> Unmanaged<CGPath>?)?,
            next: ((UnsafeRawPointer, UnsafePointer<ORBPath.CallbacksExtended>) -> UnsafePointer<ORBPath.CallbacksExtended>?)?
        ) {
            self.flags = flags
            self.retain = retain
            self.release = release
            self.apply = apply
            self.isEqual = isEqual
            self.isEmpty = isEmpty
            self.isSingleElement = isSingleElement
            self.bezierOrder = bezierOrder
            self.boundingRect = boundingRect
            self.cgPath = cgPath
            self.next = next
        }
    }
}

#endif

